File: Hard/2025-09-21-1912-Design-Movie-Rental-System/solution.js
class MovieRentingSystem {
    constructor(n, entries) {
        this.shops = new Array(n).fill(0).map(() => new Map());
        this.rentedMovies = [];
        this.rentedMap = new Map();

        for (const [shop, movie, price] of entries) {
            this.shops[shop].set(movie, price);
        }
    }

    search(movie) {
        const available = [];
        for (let shop = 0; shop < this.shops.length; shop++) {
            if (this.shops[shop].has(movie)) {
                available.push([shop, movie, this.shops[shop].get(movie)]);
            }
        }
        available.sort((a, b) => a[2] - b[2] || a[0] - b[0]);
        return available.slice(0, 5).map(([shop]) => shop);
    }

    rent(shop, movie) {
        const price = this.shops[shop].get(movie);
        this.shops[shop].delete(movie);
        this.rentedMovies.push([price, shop, movie]);
        this.rentedMap.set(`${shop}-${movie}`, price);
    }

    drop(shop, movie) {
        const price = this.rentedMap.get(`${shop}-${movie}`);
        this.rentedMap.delete(`${shop}-${movie}`);
        this.rentedMovies = this.rentedMovies.filter(([p, s, m]) => !(s === shop && m === movie));
        this.shops[shop].set(movie, price);
    }

    report() {
        const reportList = this.rentedMovies.slice();
        reportList.sort((a, b) => a[0] - b[0] || a[1] - b[1] || a[2] - b[2]);
        return reportList.slice(0, 5).map(([, shop, movie]) => [shop, movie]);
    }
}

File: Hard/2025-09-21-1912-Design-Movie-Rental-System/solution.py
import heapq
from collections import defaultdict

class MovieRentingSystem:

    def __init__(self, n: int, entries: List[List[int]]):
        self.n = n
        self.movies = defaultdict(list)
        self.rented = {}
        self.price_map = {}
        
        for shop, movie, price in entries:
            self.movies[movie].append((price, shop))
            self.price_map[(shop, movie)] = price
        
        for movie in self.movies:
            self.movies[movie].sort()

    def search(self, movie: int) -> List[int]:
        if movie not in self.movies:
            return []
        available_shops = [(price, shop) for price, shop in self.movies[movie] if (shop, movie) not in self.rented]
        return [shop for _, shop in sorted(available_shops)[:5]]

    def rent(self, shop: int, movie: int) -> None:
        price = self.price_map[(shop, movie)]
        self.rented[(shop, movie)] = price
        self.movies[movie] = [(p, s) for p, s in self.movies[movie] if s != shop]

    def drop(self, shop: int, movie: int) -> None:
        price = self.rented.pop((shop, movie))
        self.movies[movie].append((price, shop))
        self.movies[movie].sort()

    def report(self) -> List[List[int]]:
        rented_movies = [(price, shop, movie) for (shop, movie), price in self.rented.items()]
        rented_movies.sort()
        return [[shop, movie] for _, shop, movie in rented_movies[:5]]